exla/lib/exla/builder.ex

```diff
@@ -3,25 +3,75 @@ defmodule EXLA.Builder do
   Wrapper around XLA's builder.
   """
 
-  alias __MODULE__
-  alias EXLA.{Computation, Op}
+  alias EXLA.Computation
+  alias EXLA.Op
+  alias EXLA.MLIR.Module, as: M
+  alias EXLA.MLIR.Type
 
   @enforce_keys [:ref]
   defstruct [:ref, :parent, :name]
 
+  def new(name, _inputs, _outputs, :xla) do
+    new(name)
+  end
+
+  def new(name, inputs, outputs, :mlir) do
+    mlir_arg_types = mlir_type(Enum.map(inputs, &elem(&1, 1)))
+    mlir_ret_type = mlir_type(outputs)
+
+    xla_ret_shape = exla_shape(outputs)
+
+    module = M.new()
+    M.create_function(module, "main", mlir_arg_types, mlir_ret_type, xla_ret_shape)
+  end
+
+  defp mlir_type(input) when is_tuple(input) do
+    input
+    |> Tuple.to_list()
+    |> Enum.map(&mlir_type/1)
+  end
+
+  defp mlir_type(inputs) when is_list(inputs) do
+    Enum.map(inputs, &mlir_type/1)
+  end
+
+  defp mlir_type(%EXLA.Shape{} = shape) do
+    Type.new(shape)
```

Comment:
I would move the creation of Type.new fully to the C code right now. Have the C function accept EXLA.Shape and convert to Type there, get rid of the Type module. Otherwise it will be confusing keeping both `mlir_ret_type` and `xla_ret_shape`. We just need one type and the easier for now is EXLA.Shape. We can revisit later.

Reply:
Done!